File: apps/web/src/api/conduct.ts
import { api } from './client';

export interface FollowUpHistoryEntry {
  q: string;
  a?: string;
  asked_at?: string;
}

export interface QuestionConduct {
  id: string;
  order_index: number;
  question_text: string;
  question_type: 'behavioural' | 'technical' | 'coding' | 'system_design';
  round: 'R1' | 'R2' | 'R3';
  generated_by: string;
  time_minutes: number;
  expected_answer: string | null;
  evaluation_checklist: string[];
  default_follow_ups: string[];
  conduct_status: 'pending' | 'asked' | 'skipped' | 'done';
  score: number | null;
  notes: string | null;
  coverage: string[];
  follow_up_history: FollowUpHistoryEntry[];
}

export interface ConductState {
  interview_id: string;
  candidate_name: string | null;
  candidate_email: string;
  role_type: string;
  current_round: 'R1' | 'R2' | 'R3';
  experience_level: string;
  custom_instructions: string | null;
  tone: string;
  language: string;
  duration_limit_seconds: number;
  question_limit: number;
  status: string;
  questions: QuestionConduct[];
}

export interface ResponseUpsertInput {
  score?: number | null;
  notes?: string | null;
  coverage?: string[];
  follow_up_history?: FollowUpHistoryEntry[];
  conduct_status?: 'asked' | 'skipped' | 'done';
}

export interface FinalizeReport {
  interview_id: string;
  overall_score: number;
  recommendation: 'strong_hire' | 'hire' | 'no_hire' | 'strong_no_hire';
  advance_to_next_round: boolean;
  advance_reasoning: string;
  strengths: string[];
  concerns: string[];
  per_question_summary: { order_index: number; score: number; summary: string }[];
}

export async function getConductState(interviewId: string): Promise<ConductState> {
  const res = await api.get<ConductState>(`/interviews/${interviewId}/conduct`);
  return res.data;
}

export async function startConducting(interviewId: string): Promise<ConductState> {
  const res = await api.post<ConductState>(
    `/interviews/${interviewId}/conduct/start`,
    undefined,
    { timeout: 120_000 },
  );
  return res.data;
}

export async function upsertResponse(
  questionId: string,
  payload: ResponseUpsertInput,
): Promise<QuestionConduct> {
  const res = await api.patch<QuestionConduct>(`/questions/${questionId}/response`, payload);
  return res.data;
}

export async function generateFollowUp(
  questionId: string,
  payload: { notes?: string | null; coverage?: string[] },
): Promise<string> {
  const res = await api.post<{ follow_up: string }>(`/questions/${questionId}/follow-up`, payload, {
    timeout: 30_000,
  });
  return res.data.follow_up;
}

export async function finalizeInterview(interviewId: string): Promise<FinalizeReport> {
  const res = await api.post<FinalizeReport>(
    `/interviews/${interviewId}/conduct/finalize`,
    undefined,
    { timeout: 90_000 },
  );
  return res.data;
}

export interface GenerateMoreInput {
  count: number;
  focus?: string;
}

export async function generateMoreQuestions(
  interviewId: string,
  payload: GenerateMoreInput,
): Promise<ConductState> {
  const res = await api.post<ConductState>(
    `/interviews/${interviewId}/conduct/generate-more`,
    payload,
    { timeout: 120_000 },
  );
  return res.data;
}

File: apps/api/app/schemas/conduct.py
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.db.models import RoleType, Round


class QuestionConduct(BaseModel):
    """A question as seen by the recruiter while conducting."""

    id: UUID
    order_index: int
    question_text: str
    question_type: str
    round: Round
    generated_by: str
    time_minutes: int = 5

    expected_answer: str | None
    evaluation_checklist: list[str]
    default_follow_ups: list[str]
    conduct_status: str  # pending | asked | skipped | done

    # Response (nullable until first scored)
    score: int | None = None
    notes: str | None = None
    coverage: list[str] = Field(default_factory=list)
    follow_up_history: list[dict[str, Any]] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class ConductState(BaseModel):
    interview_id: UUID
    candidate_name: str | None
    candidate_email: str
    role_type: RoleType
    current_round: Round
    experience_level: str
    custom_instructions: str | None
    tone: str
    language: str
    duration_limit_seconds: int
    question_limit: int

    questions: list[QuestionConduct]
    status: str  # interview.status


class ResponseUpsert(BaseModel):
    score: int | None = Field(default=None, ge=1, le=5)
    notes: str | None = None
    coverage: list[str] = Field(default_factory=list)
    follow_up_history: list[dict[str, Any]] = Field(default_factory=list)
    conduct_status: str | None = None  # asked | skipped | done


class FollowUpRequest(BaseModel):
    notes: str | None = None
    coverage: list[str] = Field(default_factory=list)


class FollowUpResponse(BaseModel):
    follow_up: str


class GenerateMoreRequest(BaseModel):
    count: int = Field(default=5, ge=1, le=10)
    focus: str | None = None  # extra steer for THIS batch (combined with interview's custom_instructions)


class FinalizeResponse(BaseModel):
    interview_id: UUID
    overall_score: int
    recommendation: str
    advance_to_next_round: bool
    advance_reasoning: str
    strengths: list[str]
    concerns: list[str]
    per_question_summary: list[dict[str, Any]]
